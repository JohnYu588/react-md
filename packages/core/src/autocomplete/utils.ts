import { type RefObject } from "react";
import {
  type EditableHTMLElement,
  triggerManualChangeEvent,
} from "../form/utils.js";
import { type NonNullMutableRef, type UseStateInitializer } from "../types.js";
import { noopAutocompleteFilter } from "./defaults.js";
import {
  type AutocompleteFilterOptions,
  type AutocompleteGetOptionLabel,
  type AutocompleteOption,
} from "./types.js";

/**
 * @since 6.0.0
 * @internal
 */
export interface GetDefaultValueOptions<Option extends AutocompleteOption> {
  query: string | undefined;
  multiselect?: boolean;
  defaultQuery: UseStateInitializer<string> | undefined;
  defaultValue:
    | UseStateInitializer<Option | null | readonly Option[]>
    | undefined;
  options: readonly Option[];
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  filter: (options: AutocompleteFilterOptions<Option>) => readonly Option[];
}

/**
 * @since 6.0.0
 * @internal
 */
export function getDefaultValue<Option extends AutocompleteOption>(
  options: GetDefaultValueOptions<Option>
): UseStateInitializer<Option | null | readonly Option[]> {
  const {
    query,
    filter,
    multiselect,
    defaultQuery,
    defaultValue,
    options: values,
    getOptionLabel,
  } = options;

  if (typeof defaultValue !== "undefined") {
    return defaultValue;
  }

  // do not support determining the default value for a multiselect
  // autocomplete if only a query was provided.
  if (multiselect) {
    return [];
  }

  // do not support a default value out of the box for this case
  if (filter === noopAutocompleteFilter) {
    return null;
  }

  return () => {
    let q = "";
    if (query) {
      q = query;
    } else if (defaultQuery) {
      q = defaultQuery instanceof Function ? defaultQuery() : defaultQuery;
    }

    if (!q) {
      return null;
    }

    const filtered = filter({
      list: values,
      query: q,
      extractor: getOptionLabel,
    });

    return filtered[0] ?? null;
  };
}

/**
 * @since 6.0.0
 * @internal
 */
interface GetDefaultQueryOptions<Option extends AutocompleteOption> {
  value: Option | null | readonly Option[];
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  defaultQuery?: UseStateInitializer<string>;
}

/**
 * @since 6.0.0
 * @internal
 */
export function getDefaultQuery<Option extends AutocompleteOption>(
  options: GetDefaultQueryOptions<Option>
): UseStateInitializer<string> {
  const { value, getOptionLabel, defaultQuery } = options;

  if (defaultQuery) {
    return defaultQuery;
  }

  return () => {
    if (typeof value === "string") {
      return value;
    }

    if (value && typeof value === "object" && !("length" in value)) {
      return getOptionLabel(value);
    }

    return "";
  };
}

/**
 * @since 6.0.0
 * @internal
 */
export interface EnforceSelectedValueOptions<
  Option extends AutocompleteOption,
> {
  value: Option | readonly Option[] | null;
  visible: boolean;
  container: HTMLElement | null;
  popupRef: RefObject<HTMLElement>;
  comboboxRef: RefObject<EditableHTMLElement>;
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  availableOptions: readonly Option[];
  prevAvailableOptions: NonNullMutableRef<readonly Option[] | null>;
}

/**
 * This enforces that if the user clicks away, touches somewhere else on the
 * page, tabs to the next element, or focuses another element programmatically
 * the autocomplete will set the `value` back to the previous selected value if
 * there was one. i.e.
 *
 * Case 1:
 * - User selects "Apple"
 * - User hits backspace twice. Input displays "App"
 * - User clicks somewhere else on the page
 * - Input now displays "Apple" again
 *
 * Case 2:
 * - User selects "Apple"
 * - User clears the input
 * - User types "app"
 * - User clicks somewhere else on the page
 * - Input now displays "" and the value is set to `null`
 *
 * NOTE: This mutates the {@link EnforceSelectedValueOptions.prevAvailableOptions}
 *
 * @since 6.0.0
 * @internal
 */
export function enforceSelectedValue<Option extends AutocompleteOption>(
  options: EnforceSelectedValueOptions<Option>
): void {
  const {
    value,
    visible,
    container,
    popupRef,
    comboboxRef,
    getOptionLabel,
    availableOptions,
    prevAvailableOptions,
  } = options;

  if (!container) {
    return;
  }

  window.requestAnimationFrame(() => {
    if (
      container.contains(document.activeElement) ||
      popupRef.current?.contains(document.activeElement)
    ) {
      return;
    }

    if (visible) {
      // this makes it so that the options are not filtered again while closing
      prevAvailableOptions.current = availableOptions;
    }

    let label = "";
    if (typeof value === "string") {
      label = value;
    } else if (typeof value === "object" && value && !("length" in value)) {
      label = getOptionLabel(value);
    }

    triggerManualChangeEvent(comboboxRef.current, label);
  });
}
