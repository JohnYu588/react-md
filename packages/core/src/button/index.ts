export * from "./Button";
export * from "./ButtonUnstyled";
export * from "./FloatingActionButton";
