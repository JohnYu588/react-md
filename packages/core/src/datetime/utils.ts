/**
 * Since time input steps are based on seconds, this is a simple helper to
 * create the step in a human-readable form.
 *
 * @since 6.3.0
 */
export interface TimeFieldStepOptions {
  /**
   * @defaultValue `0`
   */
  seconds?: number;

  /**
   * @defaultValue `0`
   */
  minutes?: number;

  /**
   * @defaultValue `0`
   */
  hours?: number;
}

/**
 * Since time input steps are based on seconds, this is a simple helper to
 * create the step in a human-readable form.
 *
 * @example Simple Example
 * ```tsx
 * const step1 = getTimeStep({ minutes: 15 });
 * const step2 = getTimeStep({ hours: 1 });
 * const step3 = getTimeStep({ seconds: 15, minutes: 30, hours: 2 });
 * ```
 *
 * @since 6.3.0
 */
export function getTimeStep(
  step: TimeFieldStepOptions | "any" | number | undefined
): number | "any" | undefined {
  if (!step || typeof step === "string" || typeof step === "number") {
    return step;
  }

  const { hours = 0, minutes = 0, seconds = 0 } = step;

  const total = Math.abs(Math.round(seconds + minutes * 60 + hours * 60 * 60));
  return total === 0 ? undefined : total;
}
