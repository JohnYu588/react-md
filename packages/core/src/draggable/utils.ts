import type { MouseEvent, RefObject, TouchEvent } from "react";
import type { UseStateInitializer } from "../types";
import type { ClientPositionEvent, ClientPositionOptions } from "../utils";
import {
  getClientPosition,
  getRangeSteps,
  nearest,
  withinRange,
} from "../utils";

/**
 * @internal
 */
export const isMouseDragStartEvent = (event: MouseEvent): boolean =>
  event.button === 0 &&
  !event.altKey &&
  !event.metaKey &&
  !event.ctrlKey &&
  !event.shiftKey;

/**
 * @internal
 */
export const isTouchDragStartEvent = (event: TouchEvent): boolean =>
  event.changedTouches.length === 1;

/**
 * @internal
 */
interface DragPositionOptions extends ClientPositionOptions {
  isRTL: boolean;
  reversed: boolean;
  container: Element;
}

/**
 * @internal
 */
export const getDragPosition = (options: DragPositionOptions): number => {
  const { isRTL, reversed, vertical, container } = options;

  const clientPosition = getClientPosition(options);
  const { left, right, top } = container.getBoundingClientRect();
  if (vertical) {
    if (reversed) {
      return window.innerHeight - clientPosition;
    }

    // added `Math.max` since the `top` will be a negative number if rendered
    // within a portal element like a dialog/sheet
    return clientPosition - Math.max(0, top);
  }

  if (reversed ? !isRTL : isRTL) {
    return right - clientPosition;
  }

  return clientPosition - left;
};

/**
 * @internal
 */
interface RelativeDragPositionOptions extends DragPositionOptions {
  min: number;
  max: number;
  step: number;
}

/**
 * @internal
 */
interface RelativeDragPosition {
  value: number;
  dragPercentage: number;
}

/**
 * @internal
 */
export const getRelativeDragPosition = (
  options: RelativeDragPositionOptions
): RelativeDragPosition => {
  const { min, max, step, isRTL, vertical, container } = options;

  const { height, width, left, top } = container.getBoundingClientRect();
  const containerSize = vertical ? height : width;
  const containerPosition = vertical ? top + height : left;
  const clientPosition = getClientPosition(options);

  const position = vertical
    ? containerPosition - clientPosition
    : clientPosition - containerPosition;
  const distanceDragged = Math.min(Math.max(0, position), containerSize);
  let dragPercentage = distanceDragged / containerSize;
  if (isRTL && !vertical) {
    dragPercentage = 1 - dragPercentage;
  }

  const range = max - min;
  const steps = getRangeSteps({ min, max, step });
  const value = dragPercentage * range + min;
  const nextValue = nearest({
    min,
    max,
    steps,
    range,
    value,
  });

  return {
    value: nextValue,
    dragPercentage,
  };
};

/**
 * @internal
 */
interface UpdateDragPositionOptions
  extends Omit<RelativeDragPositionOptions, "container"> {
  event: ClientPositionEvent;
  nodeRef: RefObject<HTMLElement>;
  focus?: boolean;
  rangeMin: number;
  rangeMax: number;
  isDragStart: boolean;
  setValue(value: number): void;
  setDragging(dragging: boolean): void;
  setDragPercentage(value: number): void;
  withinOffsetParent: boolean;
}

/**
 * @internal
 */
export const updateDragPosition = (
  options: UpdateDragPositionOptions
): void => {
  const {
    event,
    nodeRef,
    min,
    max,
    step,
    rangeMin,
    rangeMax,
    isRTL,
    reversed,
    vertical,
    isDragStart,
    setValue,
    setDragging,
    setDragPercentage,
    withinOffsetParent,
  } = options;

  const element = nodeRef.current;
  if (!element) {
    return;
  }

  if (isDragStart) {
    // need to focus so that Chromium based browsers allow drag events. this
    // really appeared while trying to create a DnD tree
    element.focus({ preventScroll: true });

    if (!withinOffsetParent && !("changedTouches" in event)) {
      return;
    }

    // unlike the other flow, start dragging immediately so that you can trigger
    // a mousedown or touchstart event on the container element and drag until
    // the user lets go
    setDragging(true);
  }

  // firefox defaults to `document.body` while chrome will return `null`
  const container = element.offsetParent || document.body;
  if (!withinOffsetParent) {
    const dragPosition = getDragPosition({
      event,
      isRTL,
      reversed,
      vertical,
      container,
    });
    const nextValue = withinRange({
      min,
      max,
      value: dragPosition,
    });
    setValue(nextValue);

    return;
  }

  const { value, dragPercentage } = getRelativeDragPosition({
    min: rangeMin,
    max: rangeMax,
    step,
    event,
    isRTL,
    reversed,
    vertical,
    container,
  });

  setValue(value);
  setDragPercentage(dragPercentage);
};

/**
 * @internal
 */
export interface DraggableDefaultValueOptions {
  min: number;
  max: number;
  defaultValue: UseStateInitializer<number> | undefined;
}

/**
 * @internal
 */
export function getDraggableDefaultValue(
  options: DraggableDefaultValueOptions
): number {
  const { min, max, defaultValue } = options;

  if (typeof defaultValue === "function") {
    return defaultValue();
  }

  if (typeof defaultValue === "undefined") {
    return Math.ceil((max - min) / 2);
  }

  return defaultValue;
}
