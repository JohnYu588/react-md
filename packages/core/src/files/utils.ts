import { type ChangeEventHandler, type DragEventHandler } from "react";

/** @since 2.9.0 */
export interface BaseFileUploadStats {
  /**
   * A unique key associated with each upload generated by `nanoid`.
   */
  key: string;

  /**
   * The file instance that is being uploaded.
   */
  file: File;

  /**
   * The current upload progress as a percentage from 0 - 100 percent.
   */
  progress: number;
}

/** @since 2.9.0 */
export interface ProcessingFileUploadStats extends BaseFileUploadStats {
  status: "pending" | "uploading";
}

/** @since 2.9.0 */
export type FileReaderResult = FileReader["result"];

/** @since 2.9.0 */
export interface CompletedFileUploadStats extends BaseFileUploadStats {
  status: "complete";

  /**
   * The result after a `FileReader` has read a file completely.
   *
   * Note: This _should_ be an `ArrayBuffer` if the next step is to upload to a
   * server.
   *
   * @see {@link FileReaderParser}
   * @see {@link getFileParser}
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/FileReader | FileReader}
   */
  result: FileReaderResult;
}

/** @since 2.9.0 */
export type FileUploadStats =
  | ProcessingFileUploadStats
  | CompletedFileUploadStats;

/** @since 2.9.0 */
export type FileUploadStatus = FileUploadStats["status"];

/** @since 2.9.0 */
export interface FileUploadHandlers<E extends HTMLElement> {
  onDrop?: DragEventHandler<E>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

/**
 * This will first check if the mime-type of the file starts with `text/` and
 * fallback to checking a few file names or extensions that should be considered
 * text.
 *
 * This function is not guaranteed to be 100% correct and is only useful if
 * trying to generate a preview of files uploaded to the browser.
 *
 * @param file - The file to check
 * @returns `true` if the file should be considered as a text-content file.
 * @since 2.9.0
 */
export function isTextFile(file: File): boolean {
  return /\.((j|t)sx?|json|lock|hbs|ya?ml|log|txt|md)$/i.test(file.name);
}

/**
 * This will first check if the mime-type of the file starts with `text\/` and
 * fallback to checking a few file names or extensions that should be considered
 * text.
 *
 * This function is not guaranteed to be 100% correct and is only useful if
 * trying to generate a preview of files uploaded to the browser.
 *
 * @param file - The file to check
 * @returns `true` if the file should be considered as a text content file.
 * @since 2.9.0
 */
export function isImageFile(file: File): boolean {
  return /\.(a?png|avif|svg|tiff|gifv?|jpe?g)/i.test(file.name);
}

/**
 * This will first check if the mime-type of the file starts with `audio/` and
 * fallback to checking a few file names or extensions that should be considered
 * audio.
 *
 * This function is not guaranteed to be 100% correct and is only useful if
 * trying to generate a preview of files uploaded to the browser.
 *
 * @param file - The file to check
 * @returns `true` if the file should be considered as a audio content file.
 * @since 2.9.0
 */
export function isAudioFile(file: File): boolean {
  return /\.(mp3|wav|ogg|m4p|flac)$/i.test(file.name);
}

/**
 * This will first check if the mime-type of the file starts with `video/` and
 * fallback to checking a few file names or extensions that should be considered
 * video.
 *
 * This function is not guaranteed to be 100% correct and is only useful if
 * trying to generate a preview of files uploaded to the browser.
 *
 * @param file - The file to check
 * @returns `true` if the file should be considered as a video content file.
 * @since 2.9.0
 */
export function isVideoFile(file: File): boolean {
  return /\.(mkv|mpe?g|mov|avi|flv|webm|mp4)$/i.test(file.name);
}

/**
 * This function is not guaranteed to be 100% correct and is only useful if
 * trying to generate a preview of files uploaded to the browser.
 *
 * @param file - The file to check
 * @returns `true` if the file matches an image, audio, or video file.
 * @since 2.9.0
 */
export function isMediaFile(file: File): boolean {
  return isImageFile(file) || isAudioFile(file) || isVideoFile(file);
}

/**
 * One of the function names from a `FileReader` to upload a file to the
 * client.
 *
 * Note: If this file does not need to be previewed in the browser and will
 * immediately be uploaded to a server, use `readAsArrayBuffer`.
 *
 * @since 2.9.0
 * @since 6.0.0 Removed `"readAsBinaryString` since it is deprecated
 * {@see https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsBinaryString}
 */
export type FileReaderParser =
  | "readAsText"
  | "readAsDataURL"
  | "readAsArrayBuffer";

/**
 * A function that should return one of the {@link FileReaderParser} functions
 * to start uploading a file to the browser.
 *
 * @example The Default File Upload Parser
 * ```ts
 * export const getFileParser: GetFileParser = (file) => {
 *   if (isMediaFile(file)) {
 *     return "readAsDataURL";
 *   }
 *
 *   if (isTextFile(file)) {
 *     return "readAsText";
 *   }
 *
 *   return "readAsArrayBuffer";
 * };
 * ```
 *
 * @param file - The file to get a parser for
 * @returns the {@link FileReaderParser} string.
 * @since 2.9.0
 */
export type GetFileParser = (file: File) => FileReaderParser;

/**
 * This function will attempt to read:
 * - media (image, audio, and video) files as a data url so they can be
 *   previewed in `<img>`, `<audio>`, and `<video>` tags
 * - text files as plain text
 * - everything else as an `ArrayBuffer` which can be manually converted into a
 *   data url if needed with `URL.createObjectURL`
 *
 * @since 2.9.0
 */
export const getFileParser: GetFileParser = (file) => {
  if (isMediaFile(file)) {
    return "readAsDataURL";
  }

  if (isTextFile(file)) {
    return "readAsText";
  }

  return "readAsArrayBuffer";
};

/** @since 2.9.0 */
export interface SplitFileUploads {
  readonly pending: readonly ProcessingFileUploadStats[];
  readonly uploading: readonly ProcessingFileUploadStats[];
  readonly complete: readonly CompletedFileUploadStats[];
}

/**
 * This util will split all the current upload stats by status.
 *
 * @param stats - The {@link FileUploadStats} list generally returned by the
 * {@link useFileUpload} hook.
 * @returns the {@link SplitFileUploads}.
 * @since 2.9.0
 */
export function getSplitFileUploads(
  stats: readonly FileUploadStats[]
): SplitFileUploads {
  const pending: ProcessingFileUploadStats[] = [];
  const uploading: ProcessingFileUploadStats[] = [];
  const complete: CompletedFileUploadStats[] = [];
  stats.forEach((stat) => {
    if (stat.status === "pending") {
      pending.push(stat);
    } else if (stat.status === "uploading") {
      uploading.push(stat);
    } else if (stat.status === "complete") {
      complete.push(stat);
    } else {
      /* istanbul ignore next */
      throw new Error("Invalid upload stat");
    }
  });

  return { pending, uploading, complete };
}
