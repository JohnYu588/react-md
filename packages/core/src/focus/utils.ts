export const FOCUSABLE_ELEMENTS = ["BUTTON", "TEXTAREA", "SELECT"];

const BASE_FOCUSABLE_ELEMENTS =
  'a[href],area[href],input:not([disabled]):not([type="hidden"])';
const BASE_FOCUSABLE_QUERY = FOCUSABLE_ELEMENTS.reduce(
  (queryString, element) => `${queryString},${element}:not([disabled])`,
  BASE_FOCUSABLE_ELEMENTS
);

/**
 * A query selector to find elements that are programmatically focusable.
 *
 * Example:
 *
 * ```ts
 * const focusableElements = document.querySelectorAll(PROGRAMMATICALLY_FOCUSABLE);
 * // do something with elements
 * ```
 */
export const PROGRAMMATICALLY_FOCUSABLE = `${BASE_FOCUSABLE_QUERY},[tabindex]`;

/**
 * A query selector to find elements that are focusable only with tab and shift+tab.
 *
 * Example:
 *
 * ```ts
 * const focusableElements = document.querySelectorAll(TAB_FOCUSABLE);
 * // do something with elements
 * ```
 */
export const TAB_FOCUSABLE = `${PROGRAMMATICALLY_FOCUSABLE}:not([tabindex="-1"])`;

/**
 * A simple util that will find all the tab focusable elements within a
 * container element.  The container should normally be a specific HTMLElement,
 * but it can also be the entire document if you want to find **all** focusable
 * elements within your page.
 *
 * @param container - The container element/document to find focusable elements
 * within.
 * @param programmatic - Boolean if programmatically focusable elements should be
 * included instead of only tab focusable.
 * @returns A list of HTMLElements that are focusable within the container.
 */
export function getFocusableElements(
  container: HTMLElement | Document,
  programmatic = false
): readonly HTMLElement[] {
  return Array.from(
    container.querySelectorAll<HTMLElement>(
      programmatic ? PROGRAMMATICALLY_FOCUSABLE : TAB_FOCUSABLE
    )
  );
}

/**
 * An element can be tab focused if it is:
 * - an anchor or area with an `href`
 * - a non-disabled `input` element that is not `type="hidden"`
 * - a non-disabled `button`, `textarea`, or `select` element
 * - an element with a `tabIndex >= 0`
 *
 * An element can be noted as "programmatically focusable only"  has the above
 * rules, but the `tabIndex` will be set to `-1`.
 *
 * @remarks \@since 2.8.0
 */
export type ElementFocusType = "tab" | "programmatic";

export function isFocusable(
  element: HTMLElement | Document | Window,
  type: ElementFocusType = "programmatic"
): element is HTMLElement {
  const selector = type === "tab" ? TAB_FOCUSABLE : PROGRAMMATICALLY_FOCUSABLE;

  return "matches" in element && element.matches(selector);
}

export type FocusElementWithinType =
  | "first"
  | "last"
  | "query"
  | "self"
  | "none";

export interface FocusElementWithinOptions {
  type: FocusElementWithinType;
  query?: string;
  elements: readonly HTMLElement[];
  container: HTMLElement | Document;
}

export function focusElementWithin(options: FocusElementWithinOptions): void {
  const { type, query = "", container, elements } = options;
  if (type === "none") {
    return;
  }

  let element: HTMLElement | null = null;
  if (type === "first") {
    [element] = elements;
  } else if (type === "last") {
    element = elements[elements.length - 1];
  } else if (type === "query") {
    element = document.querySelector<HTMLElement>(query);
  }

  if (!element && isFocusable(container)) {
    element = container;
  }

  // TODO: Why did I add preventScroll?
  element?.focus();
}
