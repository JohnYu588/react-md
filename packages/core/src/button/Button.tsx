"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
