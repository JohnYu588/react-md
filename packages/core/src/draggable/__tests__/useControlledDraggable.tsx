import { cnb } from "cnbuilder";
import type { ReactElement } from "react";
import { useState } from "react";
import { fireEvent, rmdRender } from "../../test-utils";

import { Button } from "../../button/Button";
import { getPercentage } from "../../utils/getPercentage";
import { useControlledDraggable } from "../useControlledDraggable";

describe("useControlledDraggable", () => {
  it("should allow for the value to be controlled", () => {
    function Test(): ReactElement {
      const [value, setValue] = useState(20);
      const [dragging, setDragging] = useState(false);

      const {
        draggableRef,
        keyboardEventHandlers,
        minimum,
        maximum,
        increment,
        decrement,
      } = useControlledDraggable({
        min: 0,
        max: 100,
        dragging,
        setValue,
        setDragging,
      });

      const percentage = getPercentage({ min: 0, max: 100, value });

      return (
        <>
          <Button
            aria-valuenow={Math.ceil(percentage * 100)}
            ref={draggableRef}
            {...keyboardEventHandlers}
            className={cnb(dragging && "dragging")}
          >
            Button
          </Button>
          <Button onClick={minimum}>Minimum</Button>
          <Button onClick={maximum}>Maximum</Button>
          <Button onClick={increment}>Increment</Button>
          <Button onClick={decrement}>Decrement</Button>
        </>
      );
    }

    const { getByRole } = rmdRender(<Test />);

    const button = getByRole("button", { name: "Button" });
    const minimum = getByRole("button", { name: "Minimum" });
    const maximum = getByRole("button", { name: "Maximum" });
    const increment = getByRole("button", { name: "Increment" });
    const decrement = getByRole("button", { name: "Decrement" });
    expect(button).toMatchSnapshot();
    expect(button).toHaveAttribute("aria-valuenow", "20");

    fireEvent.click(increment);
    expect(button).toHaveAttribute("aria-valuenow", "21");

    fireEvent.click(decrement);
    expect(button).toHaveAttribute("aria-valuenow", "20");

    fireEvent.click(minimum);
    expect(button).toHaveAttribute("aria-valuenow", "0");
    fireEvent.click(maximum);
    expect(button).toHaveAttribute("aria-valuenow", "100");
  });
});
