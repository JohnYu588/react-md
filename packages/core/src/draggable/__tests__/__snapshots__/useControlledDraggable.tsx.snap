// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useControlledDraggable should allow for the value to be controlled 1`] = `
<button
  aria-valuenow="20"
  class="rmd-button rmd-button--text"
  type="button"
>
  <span>
    Button
  </span>
</button>
`;
